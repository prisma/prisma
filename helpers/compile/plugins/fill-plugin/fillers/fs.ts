export function existsSync() {
  return false
}

/**
 * A stub for fs for tryLoadEnv not to attempt loading .env
 */
const fs = {
  existsSync() {
    return false
  },
}

export default fs
