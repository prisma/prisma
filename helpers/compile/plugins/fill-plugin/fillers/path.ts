export function resolve(...args: any[]) {
  return args.join('/')
}

export function join(...args: any[]) {
  return args.join('/')
}

export const sep = '/'

export const posix = {
  sep,
}

/**
 * A poor man's shim for the "path" module
 */
const path = {
  resolve,
  posix,
  join,
  sep,
}

export default path
