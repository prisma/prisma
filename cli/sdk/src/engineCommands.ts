import { getPlatform } from '@prisma/get-platform'
import chalk from 'chalk'
import execa from 'execa'
import path from 'path'
import { ConfigMetaFormat } from './isdlToDatamodel2'
import { DMMF } from '@prisma/generator-helper'
import tmpWrite from 'temp-write'

async function getPrismaPath(): Promise<string> {
  // tslint:disable-next-line
  const dir = eval('__dirname')
  const platform = await getPlatform()
  const extension = platform === 'windows' ? '.exe' : ''
  const relative = `../query-engine-${platform}${extension}`
  return path.join(dir, relative)
}

export type GetDMMFOptions = {
  datamodel: string
  cwd?: string
  prismaPath?: string
  datamodelPath?: string
  skipRetry?: boolean
}

export async function getDMMF({
  datamodel,
  cwd = process.cwd(),
  prismaPath,
  datamodelPath,
  skipRetry,
}: GetDMMFOptions): Promise<DMMF.Document> {
  prismaPath = prismaPath || (await getPrismaPath())
  let result
  try {
    result = await execa(prismaPath, ['cli', '--dmmf'], {
      cwd,
      env: {
        ...process.env,
        PRISMA_DML: datamodel,
        PRISMA_SDL_PATH: datamodelPath,
        RUST_BACKTRACE: '1',
      },
    })

    return JSON.parse(result.stdout)
  } catch (e) {
    // If this unlikely event happens, try it at least once more
    if (e.message.includes('Command failed with exit code 26 (ETXTBSY)')) {
      await new Promise(resolve => {
        process.nextTick(resolve)
      })
      return getDMMF({
        datamodel,
        cwd,
        prismaPath,
        datamodelPath,
        skipRetry: true,
      })
    }
    if (e.stderr) {
      throw new Error(chalk.redBright.bold('Schema parsing ') + e.stderr)
    }
    if (e.message.includes('in JSON at position')) {
      throw new Error(
        `Problem while parsing the query engine response at ${prismaPath}. ${result.stdout}\n${e.stack}`,
      )
    }
    throw new Error(e)
  }
}

export async function getConfig({
  datamodel,
  cwd = process.cwd(),
  prismaPath,
  datamodelPath,
}: GetDMMFOptions): Promise<ConfigMetaFormat> {
  prismaPath = prismaPath || (await getPrismaPath())
  try {
    const result = await execa(
      prismaPath,
      ['cli', '--get_config', JSON.stringify({ datamodel }) + '\n'],
      {
        cwd,
        env: {
          ...process.env,
          PRISMA_DML: datamodel,
          PRISMA_SDL_PATH: datamodelPath,
          RUST_BACKTRACE: '1',
        },
      },
    )

    return JSON.parse(result.stdout)
  } catch (e) {
    if (e.stderr) {
      throw new Error(chalk.redBright.bold('Get config ') + e.stderr)
    }
    if (e.stdout) {
      throw new Error(chalk.redBright.bold('Get config ') + e.stdout)
    }
    throw new Error(chalk.redBright.bold('Get config ') + e)
  }
}

export interface WholeDmmf {
  dmmf: DMMF.Datamodel
  config: ConfigMetaFormat
}

export async function dmmfToDml(
  input: WholeDmmf,
  prismaPath?: string,
): Promise<string> {
  prismaPath = prismaPath || (await getPrismaPath())

  const filePath = tmpWrite.sync(JSON.stringify(input))
  try {
    const result = await execa(
      prismaPath,
      ['cli', '--dmmf_file_to_dml', filePath],
      {
        env: {
          ...process.env,
          RUST_BACKTRACE: '1',
        },
      },
    )

    return result.stdout
  } catch (e) {
    if (e.stderr) {
      throw new Error(chalk.redBright.bold('DMMF To DML ') + e.stderr)
    }
    throw new Error(e)
  }
}
