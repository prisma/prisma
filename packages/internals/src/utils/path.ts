import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
