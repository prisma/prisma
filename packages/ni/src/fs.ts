import { resolve } from 'node:path'
import fs from 'node:fs'
import type { RunnerContext } from './runner'

export function getPackageJSON(ctx?: RunnerContext): any {
  const cwd = ctx?.cwd ?? process.cwd()
  const path = resolve(cwd, 'package.json')

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw)
      return data
    }
    catch (e) {
      if (!ctx?.programmatic) {
        console.warn('Failed to parse package.json')
        process.exit(1)
      }

      throw e
    }
  }
}
