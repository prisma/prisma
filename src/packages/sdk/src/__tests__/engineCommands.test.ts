import { getDMMF, getConfig, formatSchema } from '../engineCommands'
import stripAnsi from 'strip-ansi'
import fs from 'fs'
import path from 'path'

jest.setTimeout(10000)

describe('getDMMF', () => {
  test('simple model', async () => {
    const dmmf = await getDMMF({
      datamodel: `model A {
        id Int @id
        name String
      }`,
    })

    expect(dmmf.datamodel).toMatchSnapshot()
    expect(dmmf).toMatchSnapshot()
  })

  test('@@map model', async () => {
    const dmmf = await getDMMF({
      datamodel: `
      datasource db {
        provider = "postgresql"
        url      = env("MY_POSTGRESQL_DB")
      }
      model User {
        id        Int      @default(autoincrement())
        email     String   @unique
        @@map("users")
      }`,
    })
    expect(dmmf.datamodel).toMatchSnapshot()
    expect(dmmf).toMatchSnapshot()
  })

  test('model with autoincrement should fail if sqlite', async () => {
    const datamodel = `
      datasource db {
        provider = "sqlite"
        url      = "file:dev.db"
      }
      model User {
        id        Int      @default(autoincrement())
        email     String   @unique
        @@map("users")
      }`

    /* eslint-disable jest/no-try-expect */
    try {
      await getDMMF({ datamodel })
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchSnapshot()
    }
    /* eslint-enable jest/no-try-expect */
  })

  test('model with autoincrement should fail if mysql', async () => {
    const datamodel = `
      datasource db {
        provider = "mysql"
        url      = env("MY_MYSQL_DB")
      }
      model User {
        id        Int      @default(autoincrement())
        email     String   @unique
        @@map("users")
      }`

    /* eslint-disable jest/no-try-expect */
    try {
      await getDMMF({ datamodel })
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchSnapshot()
    }
    /* eslint-enable jest/no-try-expect */
  })

  test('@@unique model', async () => {
    const dmmf = await getDMMF({
      datamodel: `
      datasource db {
        provider = "postgres"
        url      = env("MY_POSTGRES_DB")
      }

      // From https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/data-model#examples-3
      // Specify a multi-field unique attribute that includes a relation field
      model Post {
        id        Int     @default(autoincrement())
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        title     String
        published Boolean @default(false)
        
        @@unique([authorId, title])
      }
      model User {
        id    Int    @id @default(autoincrement())
        email String @unique
        posts Post[]
      }

      // Specify a multi-field unique attribute on two String fields
      model User1 {
        id        Int     @default(autoincrement())
        firstName String
        lastName  String
        isAdmin   Boolean @default(false)
        @@unique([firstName, lastName])
      }
      
      // Specify a multi-field unique attribute on two String fields and one Boolean field
      model User2 {
        id        Int     @default(autoincrement())
        firstName String
        lastName  String
        isAdmin   Boolean @default(false)
        @@unique([firstName, lastName, isAdmin])
      }
  `,
    })

    expect(dmmf).toMatchSnapshot()
  })

  test('@@unique model connectOrCreate', async () => {
    const dmmf = await getDMMF({
      datamodel: `
      datasource db {
        provider = "postgresql"
        url      = env("MY_POSTGRES_DB")
      }

      // From https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/data-model#examples-3
      // Specify a multi-field unique attribute that includes a relation field
      model Post {
        id        Int     @default(autoincrement())
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        title     String
        published Boolean @default(false)
        
        @@unique([authorId, title])
      }
      model User {
        id    Int    @id @default(autoincrement())
        email String @unique
        posts Post[]
      }

      // Specify a multi-field unique attribute on two String fields
      model User1 {
        id        Int     @default(autoincrement())
        firstName String
        lastName  String
        isAdmin   Boolean @default(false)
        @@unique([firstName, lastName])
      }
      
      // Specify a multi-field unique attribute on two String fields and one Boolean field
      model User2 {
        id        Int     @default(autoincrement())
        firstName String
        lastName  String
        isAdmin   Boolean @default(false)
        @@unique([firstName, lastName, isAdmin])
      }
  `,
      enableExperimental: ['connectOrCreate'],
    })

    expect(dmmf).toMatchSnapshot()
  })

  test('chinook introspected schema', async () => {
    const file = fs.readFileSync(
      path.join(__dirname, '../../fixtures/chinook.prisma'),
      'utf-8',
    )
    const dmmf = await getDMMF({
      datamodel: file,
    })
    const str = JSON.stringify(dmmf)
    expect(str.length).toMatchSnapshot()
  })

  test('chinook introspected schema connectOrCreate', async () => {
    const file = fs.readFileSync(
      path.join(__dirname, '../../fixtures/chinook.prisma'),
      'utf-8',
    )
    const dmmf = await getDMMF({
      datamodel: file,
      enableExperimental: ['connectOrCreate'],
    })
    const str = JSON.stringify(dmmf)
    expect(str.length).toMatchSnapshot()
  })

  test('big schema', async () => {
    const file = fs.readFileSync(
      path.join(__dirname, '../../fixtures/bigschema.prisma'),
      'utf-8',
    )
    const dmmf = await getDMMF({
      datamodel: file,
      enableExperimental: ['connectOrCreate'],
    })
    const str = JSON.stringify(dmmf)
    expect(str.length).toMatchSnapshot()
  })

  test('with validation errors', async () => {
    const datamodel = `generator client {
      provider = "prisma-client-js"
    }
    
    datasource my_db {
      provider = "sqlite"
      url      = "file:dev.db"
    }
    
    model User {
      id           String     @id @default(cuid())
      id           String     @id @default(cuid())
      name         String
      email        String     @unique
      status       String     @default("")
      permissions  Permission @default()
      permissions  Permission @default("")
      posts        Post[]
      posts        Post[]
    }
    
    model Post {
      id        String   @id @default(cuid())
      name      String
      email     String   @unique
      createdAt DateTime @default(now())
      updatedAt DateTime @updatedAt
    }
    
    enum Permission {
      ADMIN
      USER
      OWNER
      COLLABORATOR
    }
    `
    /* eslint-disable jest/no-try-expect */
    try {
      await getDMMF({ datamodel })
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchSnapshot()
    }
    /* eslint-enable jest/no-try-expect */
  })
})

describe('getConfig', () => {
  test('empty config', async () => {
    const config = await getConfig({
      datamodel: `
      datasource db {
        provider = "sqlite"
        url      = "file:../hello.db"
      }
      
      model A {
        id Int @id
        name String
      }`,
    })

    expect(config).toMatchSnapshot()
  })

  test('with generator and datasource', async () => {
    const config = await getConfig({
      datamodel: `
    datasource db {
      url = "file:dev.db"
      provider = "sqlite"
    }

    generator gen {
      provider = "fancy-provider"
      binaryTargets = ["native"]
    }

    model A {
      id Int @id
      name String
    }`,
    })

    expect(config).toMatchSnapshot()
  })

  test('datasource with env var', async () => {
    process.env.TEST_POSTGRES_URI_FOR_DATASOURCE =
      'postgres://user:password@something:5432/db'

    const config = await getConfig({
      datamodel: `
      datasource db {
        provider = "postgresql"
        url      = env("TEST_POSTGRES_URI_FOR_DATASOURCE")
      }
      `,
    })

    expect(config).toMatchSnapshot()
  })

  test('datasource with env var - ignoreEnvVarErrors', async () => {
    const config = await getConfig({
      ignoreEnvVarErrors: true,
      datamodel: `
      datasource db {
        provider = "postgresql"
        url      = env("SOMETHING-SOMETHING-1234")
      }
      `,
    })

    expect(config).toMatchSnapshot()
  })
})

describe('format', () => {
  test('nothing', async () => {
    try {
      // @ts-expect-error
      const formatted = await formatSchema({})
    } catch (e) {
      expect(e.message).toMatchSnapshot(
      )
    }
  })

  test('valid blog schemaPath', async () => {
    const formatted = await formatSchema({
      schemaPath: path.join(__dirname, 'fixtures/blog.prisma'),
    })

    expect(formatted).toMatchSnapshot()
  })

  test('valid blog schema', async () => {
    const formatted = await formatSchema({
      schema: `
      datasource db {
        provider = "sqlite"
        url      = "file:dev.db"
      }
      
      generator client {
        provider      = "prisma-client-js"
        binaryTargets = ["native"]
      }
      
      model User {
        id    String  @default(cuid()) @id
        email String  @unique
        name  String?
        posts Post[]
      }
      
      model Post {
        id        String   @default(cuid()) @id
        createdAt DateTime @default(now())
        updatedAt DateTime @updatedAt
        published Boolean
        title     String
        content   String?
        authorId  String?
        author    User?    @relation(fields: [authorId], references: [id])
      }
      
      model Like {
        id     String @default(cuid()) @id
        userId String
        user   User   @relation(fields: [userId], references: [id])
        postId String
        post   Post   @relation(fields: [postId], references: [id])
      
        @@unique([userId, postId])
      }`,
    })

    expect(formatted).toMatchSnapshot()
  })
})
