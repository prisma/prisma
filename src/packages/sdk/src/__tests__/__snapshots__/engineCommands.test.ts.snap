// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig empty config 1`] = `
Object {
  "datasources": Array [],
  "generators": Array [],
}
`;

exports[`getConfig with generator and datasource 1`] = `
Object {
  "datasources": Array [
    Object {
      "connectorType": "sqlite",
      "name": "db",
      "url": Object {
        "fromEnvVar": null,
        "value": "file:dev.db",
      },
    },
  ],
  "generators": Array [
    Object {
      "binaryTargets": Array [
        "native",
      ],
      "config": Object {},
      "name": "gen",
      "output": null,
      "provider": "fancy-provider",
    },
  ],
}
`;

exports[`getDMMF simple model 2`] = `
Object {
  "datamodel": Object {
    "enums": Array [],
    "models": Array [
      Object {
        "dbName": null,
        "fields": Array [
          Object {
            "isGenerated": false,
            "isId": true,
            "isList": false,
            "isReadOnly": false,
            "isRequired": true,
            "isUnique": false,
            "isUpdatedAt": false,
            "kind": "scalar",
            "name": "id",
            "type": "Int",
          },
          Object {
            "isGenerated": false,
            "isId": false,
            "isList": false,
            "isReadOnly": false,
            "isRequired": true,
            "isUnique": false,
            "isUpdatedAt": false,
            "kind": "scalar",
            "name": "name",
            "type": "String",
          },
        ],
        "idFields": Array [],
        "isEmbedded": false,
        "isGenerated": false,
        "name": "A",
        "uniqueFields": Array [],
      },
    ],
  },
  "mappings": Array [
    Object {
      "aggregate": "aggregateA",
      "createOne": "createOneA",
      "deleteMany": "deleteManyA",
      "deleteOne": "deleteOneA",
      "findMany": "findManyA",
      "findOne": "findOneA",
      "model": "A",
      "updateMany": "updateManyA",
      "updateOne": "updateOneA",
      "upsertOne": "upsertOneA",
    },
  ],
  "schema": Object {
    "enums": Array [
      Object {
        "name": "AOrderByInput",
        "values": Array [
          "id_ASC",
          "id_DESC",
          "name_ASC",
          "name_DESC",
        ],
      },
    ],
    "inputTypes": Array [
      Object {
        "fields": Array [
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "object",
              "type": "AWhereInput",
            },
            "name": "AND",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "object",
              "type": "AWhereInput",
            },
            "name": "OR",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "object",
              "type": "AWhereInput",
            },
            "name": "NOT",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_not",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_in",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_not_in",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_lt",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_lte",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_gt",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id_gte",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_not",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_in",
          },
          Object {
            "inputType": Object {
              "isList": true,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_not_in",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_lt",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_lte",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_gt",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_gte",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_contains",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_not_contains",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_starts_with",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_not_starts_with",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_ends_with",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name_not_ends_with",
          },
        ],
        "name": "AWhereInput",
      },
      Object {
        "fields": Array [
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id",
          },
        ],
        "name": "AWhereUniqueInput",
      },
      Object {
        "fields": Array [
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name",
          },
        ],
        "name": "ACreateInput",
      },
      Object {
        "fields": Array [
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name",
          },
        ],
        "name": "AUpdateInput",
      },
      Object {
        "fields": Array [
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "Int",
            },
            "name": "id",
          },
          Object {
            "inputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "scalar",
              "type": "String",
            },
            "name": "name",
          },
        ],
        "name": "AUpdateManyMutationInput",
      },
    ],
    "outputTypes": Array [
      Object {
        "fields": Array [
          Object {
            "args": Array [],
            "name": "id",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "Int",
            },
          },
          Object {
            "args": Array [],
            "name": "name",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "String",
            },
          },
        ],
        "name": "A",
      },
      Object {
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereInput",
                },
                "name": "where",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "enum",
                  "type": "AOrderByInput",
                },
                "name": "orderBy",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "skip",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "after",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "before",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "first",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "last",
              },
            ],
            "name": "count",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "Int",
            },
          },
        ],
        "name": "AggregateA",
      },
      Object {
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereInput",
                },
                "name": "where",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "enum",
                  "type": "AOrderByInput",
                },
                "name": "orderBy",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "skip",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "after",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "before",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "first",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Int",
                },
                "name": "last",
              },
            ],
            "name": "findManyA",
            "outputType": Object {
              "isList": true,
              "isRequired": true,
              "kind": "object",
              "type": "A",
            },
          },
          Object {
            "args": Array [],
            "name": "aggregateA",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "object",
              "type": "AggregateA",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "where",
              },
            ],
            "name": "findOneA",
            "outputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "object",
              "type": "A",
            },
          },
        ],
        "name": "Query",
      },
      Object {
        "fields": Array [
          Object {
            "args": Array [],
            "name": "count",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "Int",
            },
          },
        ],
        "name": "BatchPayload",
      },
      Object {
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "ACreateInput",
                },
                "name": "data",
              },
            ],
            "name": "createOneA",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "object",
              "type": "A",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "where",
              },
            ],
            "name": "deleteOneA",
            "outputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "object",
              "type": "A",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AUpdateInput",
                },
                "name": "data",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "where",
              },
            ],
            "name": "updateOneA",
            "outputType": Object {
              "isList": false,
              "isRequired": false,
              "kind": "object",
              "type": "A",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AWhereUniqueInput",
                },
                "name": "where",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "ACreateInput",
                },
                "name": "create",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AUpdateInput",
                },
                "name": "update",
              },
            ],
            "name": "upsertOneA",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "object",
              "type": "A",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "object",
                  "type": "AUpdateManyMutationInput",
                },
                "name": "data",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereInput",
                },
                "name": "where",
              },
            ],
            "name": "updateManyA",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "object",
              "type": "BatchPayload",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "object",
                  "type": "AWhereInput",
                },
                "name": "where",
              },
            ],
            "name": "deleteManyA",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "object",
              "type": "BatchPayload",
            },
          },
          Object {
            "args": Array [
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": true,
                  "kind": "scalar",
                  "type": "String",
                },
                "name": "query",
              },
              Object {
                "inputType": Object {
                  "isList": false,
                  "isRequired": false,
                  "kind": "scalar",
                  "type": "Json",
                },
                "name": "parameters",
              },
            ],
            "name": "executeRaw",
            "outputType": Object {
              "isList": false,
              "isRequired": true,
              "kind": "scalar",
              "type": "Json",
            },
          },
        ],
        "name": "Mutation",
      },
    ],
    "rootMutationType": "Mutation",
    "rootQueryType": "Query",
  },
}
`;

exports[`getDMMF with validation errors 1`] = `
"Schema parsing
error: Field \\"id\\" is already defined on model \\"User\\".
  -->  schema.prisma:12
   | 
11 |       id           String     @id @default(cuid())
12 |       id           String     @id @default(cuid())
   | 
error: Field \\"permissions\\" is already defined on model \\"User\\".
  -->  schema.prisma:17
   | 
16 |       permissions  Permission @default()
17 |       permissions  Permission @default(\\"\\")
   | 
error: Field \\"posts\\" is already defined on model \\"User\\".
  -->  schema.prisma:19
   | 
18 |       posts        Post[]
19 |       posts        Post[]
   | 
error: Argument \\"value\\" is missing in attribute \\"@default\\".
  -->  schema.prisma:16
   | 
15 |       status       String     @default(\\"\\")
16 |       permissions  Permission @default()
   | 
error: Expected a function value, but received string value \\"\\".
  -->  schema.prisma:17
   | 
16 |       permissions  Permission @default()
17 |       permissions  Permission @default(\\"\\")
   | 
error: Error validating: You defined the enum \`Permission\`. But the current connector does not support enums.
  -->  schema.prisma:30
   | 
29 |     
30 |     enum Permission {
31 |       ADMIN
32 |       USER
33 |       OWNER
34 |       COLLABORATOR
35 |     }
   | 

Validation Error Count: 6"
`;
